app/src/tx_metadata.c

```diff
@@ -0,0 +1,189 @@
+    1, //argument index
+    'A','d', 'd', 'r', 'e', 's', 's', 0, //expected value type
+    JSMN_STRING, //expected value json token type
+};
+
+
+//List of known compressed metadata
+typedef struct {
+    const uint8_t* metadata;
+    uint16_t metadataLength;
+} known_tx_metadata_entry_t;
+
+const known_tx_metadata_entry_t KNOWN_TX_METADATA_[] = {
+    {TX_METADATA_CREATE_ACCOUNT, sizeof(TX_METADATA_CREATE_ACCOUNT)},
+    {TX_METADATA_ADD_NEW_KEY, sizeof(TX_METADATA_ADD_NEW_KEY)},
+    {TX_METADATA_TOKEN_TRANSFER, sizeof(TX_METADATA_TOKEN_TRANSFER)},
+    // sentinel, do not remove
+    {NULL, 0}
+};
+
+parser_error_t _validateHash(uint8_t scriptHash[SCRIPT_HASH_SIZE], const NV_VOLATILE uint8_t *txMetadata, uint16_t txMetadataLength) {
+    if (txMetadataLength < 1) {
+        return PARSER_TEMPLATE_ERROR;
+    }
+    uint8_t numberOfHashes = txMetadata[0];
+    if (numberOfHashes > MAX_TEMPLATE_NUMBER_OF_HASHES || txMetadataLength < 1+SCRIPT_HASH_SIZE*numberOfHashes) {
+        return PARSER_TEMPLATE_ERROR;
+    }
+
+    for(size_t i=0; i<numberOfHashes; i++) {
+        uint8_t thisHashMatches = 1;
+        for(int j=0; j<SCRIPT_HASH_SIZE; j++) {
+            uint8_t hashByte = txMetadata[1+i*SCRIPT_HASH_SIZE+j];
+            if (hashByte != scriptHash[j]) {
+                thisHashMatches = 0;
+                break;
+            }
+        }
+        if (thisHashMatches == 1) {
+            return PARSER_OK;
+        }
+    }
+
+    return PARSER_UNEXPECTED_SCRIPT;
+}
+
+
+parser_error_t parseTxMetadata(uint8_t scriptHash[SCRIPT_HASH_SIZE], const NV_VOLATILE uint8_t *txMetadata, uint16_t txMetadataLength, 
+                               parsed_tx_metadata_t *parsedTxMetadata) {
+    uint16_t parsed = 0;
+        
+    #define READ_CHAR(where) \
+        { \
+            if (!(parsed < txMetadataLength)) { \
+                return PARSER_TEMPLATE_ERROR; \
+            } \
+            (*where) = txMetadata[parsed++]; \
+        } 
+    #define READ_STRING(dest_pointer, len) { \
+        *(len) = 0; \
+        *(dest_pointer) = (char *) txMetadata + parsed; \
+        while (*(len) <= MAX_TEMPLATE_STRING_LENGTH) { \
+            uint8_t byte = 0; \
+            READ_CHAR(&byte); \
+            if (byte == 0) { \
+                break; \
+            } \
+            (*(len))++; \
+        } \
+        if (*(len) > MAX_TEMPLATE_STRING_LENGTH) { \
+            return PARSER_TEMPLATE_ERROR; \
+        } \
+    }
+    #define READ_SKIP(count) { \
+        parsed += (count); \
+    }
+
+    //read number of hashes and validate script
+    {
+        uint8_t numberOfHashes = 0;
+        READ_CHAR(&numberOfHashes)
+        if (numberOfHashes > MAX_TEMPLATE_NUMBER_OF_HASHES) {
+            return PARSER_TEMPLATE_TOO_MANY_HASHES;
+        }
+        parser_error_t err = _validateHash(scriptHash, txMetadata, txMetadataLength);
+        if (err != PARSER_OK) {
+            return err;
+        }
+        READ_SKIP(numberOfHashes*SCRIPT_HASH_SIZE)
```

Comment:
```suggestion
        READ_SKIP(numberOfHashes * SCRIPT_HASH_SIZE)
```
I think it would be more consistent with the code base and would less resemble pointer stuff.